### ISSUE TITLE
Support for IRCv3 message tags

### ISSUE THREAD

**Issue by sorcix** [sorcix/irc#3]

**Comment by MorpheusXAUT** [comment]
I've forked your library and added support for IRCv3 message tags in these two commits: https://github.com/MorpheusXAUT/irc/commit/06f5b7face4158d43b822e67eafd8751c077c6d2 & https://github.com/MorpheusXAUT/irc/commit/1aec38aa8e3b0f97db85f04fa764bfcf228294c1

Been trying the forked library + implemented tags in another chat bot project of mine since then - not seen any errors or problems so far. I didn't come around to implement unit tests yet, so that would still be missing.

Would you be interested in me trying to set up some unit tests and then sending in a pull request?

**Comment by sorcix** [comment]
Interesting. Except for the escaping, most of your changes were available in the "tags" branch already. I'll have to compare your changes and see if I can merge them in there.

**Comment by MorpheusXAUT** [comment]
Unfortunately, I realised exactly that just as I posted my comment on this issue, maybe next time reading the open issues/branches beforehand will save me some time!

Let me know if I can be of assistance.

**Comment by sorcix** [comment]
Let's compare our work and find out why we did things differently.

Looks like we both added Tags as an extra type. I decided to expose the map as a way to get access to the raw tag values (without escaping) and have a Get and Set function to do the escaping. I don't know if making it a struct with a single member has any benefits?

(I was also planning to loop over the characters to escape, instead of using the Replace method a few times. But since code is written already I guess there it doesn't hurt to merge that in already..)

**Comment by MorpheusXAUT** [comment]
To be honest, there probably is no additional use in defining an own struct to store the tags, a map would have sufficed. I figured there might be additional information to store separately when starting to implement it, but the tags turned out way easier to expected, so the struct stayed empty but the map.

As for the looping/iterating over chars: I honestly didn't have highest performance possible in mind, but used a simple string replacement assuming it would be efficient enough. We could test how large/if even there is a difference and a loop to replace escaped values would be more suitable.

An extra type defined as a map of strings would probably the best solution, just leaving it empty if no tags have been passed along with the message.

**Comment by sorcix** [comment]
Now that I know someone is using IRCv3 tags, lets get this merged in. I should have some time Wednesday to write some tests. Escaping the tag values when going from struct to raw IRC message also needs some love.

**Comment by MorpheusXAUT** [comment]
Thanks, greatly appreciated! Always prefer to use the actual library and not a quick fork by me :smile: 
Let me know if I can help you with any of the implementation needed.

**Comment by sorcix** [comment]
I didn't have as much time as I had hoped. Managed to start with the tests. To be continued.

**Comment by sorcix** [comment]
Tag escaping and unescaping is now in the [tags branch](https://github.com/sorcix/irc/tree/tags). Could you review the changes? I think there was an error in your escaping, but I could have misread the documentation on [message tags](http://ircv3.net/specs/core/message-tags-3.2.html).

**Comment by MorpheusXAUT** [comment]
Looks good to me! I'll be happy to try it out "live" in the next days and report back if I find any errors.

And yes, I indeed have an error in my tag escaping (copy/paste error for "\", incorrect replacement of CR and LFs, oops :smile: )

**Comment by sorcix** [comment]
I still need to collect some test cases and add benchmarks before it can be
merged to master. If you have any 'weird' examples, feel free to leave them
here!

Note to self: https://golang.org/pkg/strings/#Replacer

**Comment by jakebailey** [comment]
I'd like to use this as well (I've been using @MorpheusXAUT's fork, though probably not now that I see that CR LF issue, unless I fork it too). Is this still being worked on? As-is, this branch doesn't actually parse the tags on real messages anyway. There doesn't seem to be any Go IRC libraries that parse tags (and this is the library I really want to use).

**Comment by MorpheusXAUT** [comment]
@jakebailey I've been using the tags code from my branch without problems basically ever since I've made the last commits on there, "live-testing" it with the Twitch IRC servers and multiple channels with several hundred to thousand different chatters daily - haven't seen any issues so far :smile: 

I'm happy to merge in any additional commits as well (or maybe look into the CR LF issue myself if I find the time in the next days). Since I've seen you make some commit as well, I'd also be happy to look into them/additional tests if you want.

**Comment by jakebailey** [comment]
I just did a quick fork of your code and changed it a bit in the meantime. That'll work for me (I'll just import from my github), but if you want, I could PR you. It wouldn't really affect me, but maybe help if you have any users.

Personally, I would not make a struct, and just make a type directly from the map, like @sorcix did. I'll probably do that later today on my fork, in addition to some of the fixes I did on your code.

**Comment by MorpheusXAUT** [comment]
Fair enough, yeah, a struct wasn't really needed and I planned to switch it out at _some_ point, haven't gotten around to it yet though. A map definitely is enough!
I'd appreciate a PR if you find the time (or PR @sorcix directly, since I only forked it to implement this). Glad to see someone else is using this as well and (some) of my code helped out to start :smile:

**Comment by sorcix** [comment]
Oh, looks like I forgot to commit my last change. Oops. That explains why
nobody gave me any feedback. I'll check that out when I get back to my pc.
Message tags should go to master once that code has enough tests.

-- Vic

**Comment by jakebailey** [comment]
This is half-related, but if the tag code is in a seperate file (with its own tests in its own file), then should the Prefix code also do the same?

**Comment by sorcix** [comment]
The missing commit is the change in ParseMessage to send the first part to ParseTags and the similar code in message.Bytes(). I'm not sure where it is, as I use a few computers to write code.

However, looks like your branch is used already and thus better 'tested' than mine, so it may make sense to merge in your code instead. The only difference seems to be my code does the replacing only when someone asks for the tag contents, while yours does it upfront. Does not make much difference.
